File: python/semantic_kernel/agents/runtime/core/telemetry/tracing.py
# Copyright (c) Microsoft. All rights reserved.

import contextlib
from collections.abc import Iterator
from typing import Generic

from opentelemetry.trace import NoOpTracerProvider, Span, SpanKind, TracerProvider, get_tracer_provider
from opentelemetry.util import types

from semantic_kernel.agents.runtime.core.telemetry.propagation import TelemetryMetadataContainer, get_telemetry_links
from semantic_kernel.agents.runtime.core.telemetry.tracing_config import (
    Destination,
    ExtraAttributes,
    Operation,
    TracingConfig,
)
from semantic_kernel.utils.feature_stage_decorator import experimental


@experimental
class TraceHelper(Generic[Operation, Destination, ExtraAttributes]):
    """TraceHelper is a utility class to assist with tracing operations using OpenTelemetry.

    This class provides a context manager `trace_block` to create and manage spans for tracing operations,
    following semantic conventions and supporting nested spans through metadata contexts.

    """

    def __init__(
        self,
        tracer_provider: TracerProvider | None,
        instrumentation_builder_config: TracingConfig[Operation, Destination, ExtraAttributes],
    ) -> None:
        """Initialize the TraceHelper with a tracer provider and instrumentation builder config."""
        # Evaluate in order: first try tracer_provider param, then get_tracer_provider(), finally fallback to NoOp
        # This allows for nested tracing with a default tracer provided by the user
        self.tracer_provider = tracer_provider or get_tracer_provider() or NoOpTracerProvider()
        self.tracer = self.tracer_provider.get_tracer(f"agent_runtime {instrumentation_builder_config.name}")
        self.instrumentation_builder_config = instrumentation_builder_config

    @contextlib.contextmanager
    def trace_block(
        self,
        operation: Operation,
        destination: Destination,
        parent: TelemetryMetadataContainer | None,
        *,
        extraAttributes: ExtraAttributes | None = None,
        kind: SpanKind | None = None,
        attributes: types.Attributes | None = None,
        start_time: int | None = None,
        record_exception: bool = True,
        set_status_on_exception: bool = True,
        end_on_exit: bool = True,
    ) -> Iterator[Span]:
        """Thin wrapper on top of start_as_current_span.

        1. It helps us follow semantic conventions
        2. It helps us get contexts from metadata so we can get nested spans

        Args:
            operation (MessagingOperation): The messaging operation being performed.
            destination (MessagingDestination): The messaging destination being used.
            parent (Optional[TelemetryMetadataContainer]): The parent telemetry metadata context
            kind (SpanKind, optional): The kind of span. If not provided, it maps to PRODUCER or CONSUMER depending
                on the operation.
            extraAttributes (ExtraAttributes, optional): Additional defined attributes for the span. Defaults to None.
            attributes (Optional[types.Attributes], optional): Additional non-defined attributes for the span.
                Defaults to None.
            start_time (Optional[int], optional): The start time of the span. Defaults to None.
            record_exception (bool, optional): Whether to record exceptions. Defaults to True.
            set_status_on_exception (bool, optional): Whether to set the status on exception. Defaults to True.
            end_on_exit (bool, optional): Whether to end the span on exit. Defaults to True.

        Yields:
            Iterator[Span]: The span object.

        """
        span_name = self.instrumentation_builder_config.get_span_name(operation, destination)
        span_kind = kind or self.instrumentation_builder_config.get_span_kind(operation)
        context = None  # TODO(evmattso): we may need to remove other code for using custom context.
        links = get_telemetry_links(parent) if parent else None
        attributes_with_defaults: dict[str, types.AttributeValue] = {}
        for key, value in (attributes or {}).items():
            attributes_with_defaults[key] = value
        instrumentation_attributes = self.instrumentation_builder_config.build_attributes(
            operation, destination, extraAttributes
        )
        for key, value in instrumentation_attributes.items():
            attributes_with_defaults[key] = value
        with self.tracer.start_as_current_span(
            span_name,
            context,
            span_kind,
            attributes_with_defaults,
            links,
            start_time,
            record_exception,
            set_status_on_exception,
            end_on_exit,
        ) as span:
            yield span

File: nodejs/semantic-kernel/agents/runtime/core/telemetry/tracing.ts
import { Attributes, context, Span, SpanKind, SpanOptions, trace, Tracer, TracerProvider } from '@opentelemetry/api'
import { experimental } from '../../../../utils/feature-stage-decorator'
import type { TelemetryMetadataContainer } from './propagation'
import { getTelemetryContext, getTelemetryLinks } from './propagation'
import type { TracingConfig } from './tracing-config'

/**
 * TraceHelper is a utility class to assist with tracing operations using OpenTelemetry.
 *
 * This class provides methods to create and manage spans for tracing operations,
 * following semantic conventions and supporting nested spans through metadata contexts.
 */
@experimental
export class TraceHelper<TOperation extends string, TDestination, TExtraAttributes extends Record<string, any>> {
  private tracerProvider: TracerProvider
  private tracer: Tracer
  private instrumentationBuilderConfig: TracingConfig<TOperation, TDestination, TExtraAttributes>

  constructor(
    tracerProvider: TracerProvider | null | undefined,
    instrumentationBuilderConfig: TracingConfig<TOperation, TDestination, TExtraAttributes>
  ) {
    // Evaluate in order: first try tracerProvider param, then get global tracer, finally fallback to NoOp
    this.tracerProvider = tracerProvider || trace.getTracerProvider()
    this.tracer = this.tracerProvider.getTracer(`agent_runtime ${instrumentationBuilderConfig.name}`)
    this.instrumentationBuilderConfig = instrumentationBuilderConfig
  }

  /**
   * Execute a function within a traced span.
   *
   * This is a thin wrapper that helps us:
   * 1. Follow semantic conventions
   * 2. Get contexts from metadata so we can get nested spans
   *
   * @param operation - The messaging operation being performed.
   * @param destination - The messaging destination being used.
   * @param parent - The parent telemetry metadata context.
   * @param fn - The function to execute within the span.
   * @param options - Additional span options.
   * @returns The result of the function execution.
   */
  async traceBlock<T>(
    operation: TOperation,
    destination: TDestination,
    parent: TelemetryMetadataContainer,
    fn: (span: Span) => Promise<T>,
    options?: {
      extraAttributes?: TExtraAttributes
      kind?: SpanKind
      attributes?: Attributes
      startTime?: number
      recordException?: boolean
      setStatusOnException?: boolean
      endOnExit?: boolean
    }
  ): Promise<T> {
    const spanName = this.instrumentationBuilderConfig.getSpanName(operation, destination)
    const spanKind = options?.kind || this.instrumentationBuilderConfig.getSpanKind(operation)
    const parentContext = parent ? getTelemetryContext(parent) : context.active()
    const links = parent ? getTelemetryLinks(parent) : undefined

    const attributesWithDefaults: Attributes = { ...(options?.attributes || {}) }
    const instrumentationAttributes = this.instrumentationBuilderConfig.buildAttributes(
      operation,
      destination,
      options?.extraAttributes
    )

    Object.assign(attributesWithDefaults, instrumentationAttributes)

    const spanOptions: SpanOptions = {
      kind: spanKind,
      attributes: attributesWithDefaults,
      links,
      startTime: options?.startTime,
    }

    const span = this.tracer.startSpan(spanName, spanOptions, parentContext)

    try {
      return await context.with(trace.setSpan(parentContext, span), async () => {
        return await fn(span)
      })
    } catch (error) {
      if (options?.recordException !== false) {
        span.recordException(error as Error)
      }
      if (options?.setStatusOnException !== false) {
        span.setStatus({
          code: 2, // ERROR
          message: (error as Error).message,
        })
      }
      throw error
    } finally {
      if (options?.endOnExit !== false) {
        span.end(options?.startTime ? Date.now() : undefined)
      }
    }
  }

  /**
   * Synchronous version of traceBlock for non-async operations.
   */
  traceBlockSync<T>(
    operation: TOperation,
    destination: TDestination,
    parent: TelemetryMetadataContainer,
    fn: (span: Span) => T,
    options?: {
      extraAttributes?: TExtraAttributes
      kind?: SpanKind
      attributes?: Attributes
      startTime?: number
      recordException?: boolean
      setStatusOnException?: boolean
      endOnExit?: boolean
    }
  ): T {
    const spanName = this.instrumentationBuilderConfig.getSpanName(operation, destination)
    const spanKind = options?.kind || this.instrumentationBuilderConfig.getSpanKind(operation)
    const parentContext = parent ? getTelemetryContext(parent) : context.active()
    const links = parent ? getTelemetryLinks(parent) : undefined

    const attributesWithDefaults: Attributes = { ...(options?.attributes || {}) }
    const instrumentationAttributes = this.instrumentationBuilderConfig.buildAttributes(
      operation,
      destination,
      options?.extraAttributes
    )

    Object.assign(attributesWithDefaults, instrumentationAttributes)

    const spanOptions: SpanOptions = {
      kind: spanKind,
      attributes: attributesWithDefaults,
      links,
      startTime: options?.startTime,
    }

    const span = this.tracer.startSpan(spanName, spanOptions, parentContext)

    try {
      return context.with(trace.setSpan(parentContext, span), () => {
        return fn(span)
      })
    } catch (error) {
      if (options?.recordException !== false) {
        span.recordException(error as Error)
      }
      if (options?.setStatusOnException !== false) {
        span.setStatus({
          code: 2, // ERROR
          message: (error as Error).message,
        })
      }
      throw error
    } finally {
      if (options?.endOnExit !== false) {
        span.end(options?.startTime ? Date.now() : undefined)
      }
    }
  }
}
